### PR TITLE
(WIP) Support property wrappers in top-level code

### PR DESCRIPTION
**Pull request by angela-laar**

### CODE REVIEW

File: include/swift/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -2097,7 +2097,7 @@ class SolutionApplicationTarget {
     // local variables with initializers.
     bool hasInit = expression.propertyWrapper.hasInitialWrappedValue;
     if (wrappedVar->isStatic() ||
-        (hasInit && wrappedVar->getDeclContext()->isLocalContext()))
+        (hasInit && wrappedVar->getDeclContext()->isLocalContext()) || (hasInit && wrappedVar->isTopLevelGlobal()))
```
Comment (kavon) [inline include/swift/Sema/ConstraintSystem.h:2100]: could do `hasInit && (X || Y)` to make it shorter

File: lib/SIL/IR/SILDeclRef.cpp [hunk 2]
```diff
@@ -376,11 +376,8 @@ SILLinkage SILDeclRef::getLinkage(ForDefinition_t forDefinition) const {
     if (isSerialized())
       return maybeAddExternal(SILLinkage::PublicNonABI);
 
-    d = cast<NominalTypeDecl>(d->getDeclContext());
+    limit = Limit::NeverPublic;
 
-    // FIXME: This should always be true.
-    if (d->getModuleContext()->isResilient())
```
Comment (kavon) [inline lib/SIL/IR/SILDeclRef.cpp:382]: Let's convert this into an assertion to be sure. There's probably a way to get the module context of the original `d` that doesn't require it to be a NominalTypeDecl?

File: lib/SILGen/SILGen.cpp [hunk 3]
```diff
@@ -1763,6 +1763,35 @@ void SILGenModule::visitPatternBindingDecl(PatternBindingDecl *pd) {
...
+          addGlobalVariable(var);
+
+        var->visitEmittedAccessors([&](AccessorDecl *accessor){
+          emitFunction(accessor);
+        });
+
+        /*auto *asd = dyn_cast<AbstractStorageDecl>(var);
+        asd->visitEmittedAccessors([&](AccessorDecl *accessor) {
+          visitFuncDecl(accessor);
+        });*/
```
Comment (kavon) [inline lib/SILGen/SILGen.cpp:1792]: Some extra stuff you can remove.
```suggestion
```

**Review by kavon — COMMENTED** [review]
Good progress!

**Comment by angela-laar** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
Let's do this work on `main`, because Swift 5.7 is mostly taking only bug fixes at this point. I'll close this PR, but please open up another against `main` when you're ready!

**Pull request closed without merging**